miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/StorageLocationRestController.java

```diff
@@ -0,0 +1,200 @@
+  @ResponseStatus(HttpStatus.CREATED)
+  public @ResponseBody StorageLocationDto createFreezer(@RequestBody StorageLocationDto dto) {
+    return doSave(dto);
+  }
+
+  @RequestMapping(value = "/freezers/{id}", method = RequestMethod.PUT)
+  public @ResponseBody StorageLocationDto update(@PathVariable(name = "id", required = true) long id, @RequestBody StorageLocationDto dto)
+      throws IOException {
+    if (dto.getId() != id) {
+      throw new RestException("Location ID mismatch", Status.BAD_REQUEST);
+    }
+    getFreezer(id); // checks existing freezer
+    return doSave(dto);
+  }
+
+  private StorageLocationDto doSave(StorageLocationDto dto) {
+    StorageLocation freezer = Dtos.to(dto);
+    long savedId = storageLocationService.saveFreezer(freezer);
+    StorageLocation saved = storageLocationService.get(savedId);
+    return Dtos.asDto(saved, true, true);
+  }
+
+  @RequestMapping(value = "/freezers/{id}/shelves", method = RequestMethod.POST)
+  public @ResponseBody StorageLocationDto addFreezerShelf(@PathVariable(name = "id", required = true) long id) {
+    StorageLocation freezer = getFreezer(id);
+    int lastShelf = freezer.getChildLocations().stream().filter(loc -> {
+      return loc.getLocationUnit() == LocationUnit.SHELF;
+    }).map(StorageLocation::getAlias).mapToInt(Integer::parseInt).max().orElse(0);
+
+    StorageLocation shelf = new StorageLocation();
+    shelf.setAlias(Integer.toString(lastShelf + 1));
+    shelf.setLocationUnit(LocationUnit.SHELF);
+    shelf.setParentLocation(freezer);
+
+    long savedId = storageLocationService.addFreezerStorage(shelf);
+    StorageLocation saved = storageLocationService.get(savedId);
+    return Dtos.asDto(saved, false, false);
+  }
+
+  private StorageLocation getFreezer(long id) {
+    StorageLocation freezer = storageLocationService.get(id);
+    if (freezer == null || freezer.getLocationUnit() != LocationUnit.FREEZER) {
+      throw new RestException("Freezer not found", Status.NOT_FOUND);
+    }
+    return freezer;
+  }
+
+  @RequestMapping(value = "/freezers/{id}/stacks", method = RequestMethod.POST)
+  public @ResponseBody StorageLocationDto addFreezerStack(@PathVariable(name = "id", required = true) long id,
+      @RequestParam(name = "height", required = true) int height) {
+    StorageLocation freezer = getFreezer(id);
+    if (height < 1) {
+      throw new RestException("Invalid stack height", Status.BAD_REQUEST);
+    }
+
+    return createStack(freezer, height);
+  }
+
+  private StorageLocation makeStorage(String alias, LocationUnit locationUnit, StorageLocation parent) {
+    StorageLocation storage = new StorageLocation();
+    storage.setAlias(alias);
+    storage.setLocationUnit(locationUnit);
+    storage.setParentLocation(parent);
+    return storage;
+  }
+
+  private String findNextNumber(StorageLocation parent, LocationUnit childType) {
+    return Integer.toString(parent.getChildLocations().stream().filter(loc -> {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove useless curly braces around statement and then remove useless return keyword [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1602)